src/language-js/printer-estree.js

```diff
@@ -3734,32 +3735,47 @@ function printPropertyKey(path, options, print) {
         !prop.computed &&
         prop.key &&
         isStringLiteral(prop.key) &&
-        !isStringPropSafeToCoerceToIdentifier(prop, options)
+        !isStringPropSafeToUnquote(prop, options)
     );
     needsQuoteProps.set(parent, objectHasStringProp);
   }
 
   if (
-    key.type === "Identifier" &&
+    (key.type === "Identifier" ||
+      ((isNumericLiteral(key) || key.type === "BigIntLiteral") &&
+        !(options.parser === "typescript" || options.parser === "babel-ts"))) &&
     (options.parser === "json" ||
       (options.quoteProps === "consistent" && needsQuoteProps.get(parent)))
   ) {
     // a -> "a"
-    const prop = printString(JSON.stringify(key.name), options);
+    // 1e2 -> "100"
+    // 2n -> "2"
+    const prop = printString(
+      JSON.stringify(
+        key.type === "Identifier" ? key.name : key.value.toString()
+      ),
+      options
+    );
```

Comment:
Wow. Surprisingly, this behavior is correct:

**Prettier pr-8508**
[Playground link](https://deploy-preview-8508--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEB5ARgKzmGA6AazgE8BnACmAHIBzAWiqQBYAaABgA90B9dJADn7suvXlCQBmKSwA6UAATzOAQ3ETZCpegCMbPfoNtuhk4e3bxTAL4BKOSBYgIABxgBLaKWShlAJ18QAO4ACn4IXijKADaBymQOIOi+ymBEMADKziluUDTIML4ArnCOcAC26HAAJlXVADKqNIXKNHAAYhC+Zcow7rnIIMqFMBAJABYwZVEA6mNu8KRZYHDp4fNuAG7zxANgpF6OOaRwvjDByTTdyABm0ceOmKQcAELJqXAZymVwdTlwN3cSiBHhx0jkaFE4ABFQoQeAAqL3EBZXzHXwDdDKSpRBLOXw5GDTNxVGBjZD8NiOPEQY7TZLOAZ4uBojb-RwAR1h8HOLgi4E8blI8EQjl8cE5bjF5xaVyQt0RQOOZTc+SKivBkJhcP+csBjhgWKJJLJSAATPrkm4ouCAMIQMqykDMgCsCUKxwAKliIvKkRtigBJKC1WDpMD41wAQWD6RgxEhCOOVisQA)
```sh
--parser babel
--quote-props consistent
```

**Input:**
```jsx
Object.keys({'g-':4,0xb_b:88,0xb_b_bn:333,
  0xan:3,
  0b100000000000_000000000000000011n:4})

```

**Output:**
```jsx
Object.keys({
  "g-": 4,
  "187": 88,
  "0xbbb": 333,
  "0xa": 3,
  "0b100000000000000000000000000011": 4,
});

```

Reply:
Please add this to the tests.

Reply:
You mean this is a bug, right? Because these are the expected keys:

```
> Object.keys({'g-':4,0xb_b:88,0xb_b_bn:333,0xan:3,0b100000000000_000000000000000011n:4})
[ '10', '187', '3003', '536870915', 'g-' ]
```

Reply:
Fixed and test added.

Reply:
Where did you copy this from? In my Chrome 83.0.4103.97:
```console
Object.keys({'g-':4,0xb_b:88,0xb_b_bn:333,0xan:3,0b100000000000_000000000000000011n:4})
(5) ["187", "g-", "0xbbb", "0xa", "0b100000000000000000000000000011"]
```

Reply:
Node.js 14.3.0 and Firefox 77. Chrome gives the same result as for you on my machine as well.

Reply:
I believe this is a bug in Chrome. It knows how to do `.toString()`, but does not seem to use it for object keys:

```
> 0b10n.toString()
"2"
> ({0b10n: 1})
{0b10: 1}
```

Reply:
I asked on Twitter: https://twitter.com/SimonLydell/status/1269565993652715520

Reply:
I can't find anything about bigint keys in the spec: https://tc39.es/ecma262/#prod-LiteralPropertyName

Reply:
So, I read the spec. Definitely a bug in Chrome.